### ISSUE TITLE
Deeplinking with modals

### ISSUE THREAD

**Issue by Sephaq** [react-navigation/react-navigation#9276]
**Current Behavior**

When opening a deeplink URL that involves a modal the app has no stack to go back to, even though all of them have an initial route.

Profile settings is the only one that works but as you can see the name of the STACK and of the SCREEN are the same.
aside from that, inventory and message both navigate to the correct screen, but i can't navigate back since there is no screen in the stack to handle it.

**Expected Behavior**

When the user uses any deep-link url it navigates to the correct screen, and if it's a modal it has the appStack as the root of the stack.

**How to reproduce**

I'm running the following configuration for my linking:
```
const linkingConfig = {
  screens: {
    profileSettings: {
      initialRouteName:'appStack',
      screens: {
        profileSettings: 'settings'
      }
    },
    inventoryModalNav: {
      initialRouteName:'appStack',
      screens: {
        inventoryItem: 'create',
      }
    },
    messageModalNav: {
      initialRouteName:'appStack',
      screens: {
        messageDetail: 'messages/view/:messageId',
      }
    },
    appStack: {
      screens: {
        tasksStack: {
          path: 'tasks',
          screens: {
            taskDetails: 'view/:id'
          }
        },
        inventoryStack: 'inventory',
        messagesStack: 'messages'
      }
    }
  },
}
```

The followings are configured as modals:
- messageModalNav
- inventoryModalNav
- profileSettings

*Navigators*
- Root - StackNavigator
```
<RootStack.Navigator
    initialRouteName={routeName}
    screenOptions={Platform.OS === 'android' ? androidStackConfig : stackConfig}
    headerMode={"none"}
    mode="modal"
  >
    <RootStack.Screen name={RouteIdentifiers.loggedOutStack.name} component={LoggedOutStackNavigator} initialParams={initialParams}/>
    <RootStack.Screen name={RouteIdentifiers.onBoardingStack.name} component={OnBoardingStackNavigator} initialParams={initialParams}/>
    <RootStack.Screen name={RouteIdentifiers.appStack.name} component={Tabs} initialParams={initialParams}/>
    <RootStack.Screen name={RouteIdentifiers.inventoryModalNavigator.name} component={InventoryModalNavigator}/>
    <RootStack.Screen name={RouteIdentifiers.messageDetailModal.name} component={MessageModalNavigator}/>
    <RootStack.Screen name={RouteIdentifiers.sharedTaskStack.name} component={SharedTaskStackNavigator}/>
    <RootStack.Screen name={RouteIdentifiers.profileSettings.name} component={ProfileSettingsModalNavigator}/>
    <RootStack.Screen name={RouteIdentifiers.weather.name} component={WeatherNavigatorModal}/>
    <RootStack.Screen name={RouteIdentifiers.photoRequestModal.name} component={PhotoRequestModalNavigator}/>
  </RootStack.Navigator>
```
- AppStack - StackNavigator

```
    <ApplicationTabs.Navigator
      tabBar={props => <BottomNavigationTab {...props} />}
      tabBarOptions={{keyboardHidesTabBar: true}}
      initialRouteName={RouteIdentifiers.mainDashboardStack.name}
    >
      <ApplicationTabs.Screen name={RouteIdentifiers.tasksStack.name} component={TaskStackNavigator} options={({ route }) => ({tabBarLabel: strings.tabActivities, tabBarVisible: showTabBar(route) })}/>
      <ApplicationTabs.Screen name={RouteIdentifiers.messagesStack.name} component={MessagesStackNavigator} options={({ route }) => ({tabBarLabel: strings.tabMessages, tabBarVisible: showTabBar(route) })}/>
      <ApplicationTabs.Screen name={RouteIdentifiers.mainDashboardStack.name} component={MainDashboardStackNavigator} options={({ route }) => ({tabBarLabel: strings.tabMainDashboard, tabBarVisible: showTabBar(route) })}/>
      <ApplicationTabs.Screen name={RouteIdentifiers.inventoryStack.name} component={InventoryStackNavigator} options={({ route }) => ({tabBarLabel: strings.tabInventory, tabBarVisible: showTabBar(route) })}/>
      <ApplicationTabs.Screen name={RouteIdentifiers.smartDevicesStack.name} component={SmartDevicesStackNavigator} options={({ route }) => ({tabBarLabel: strings.tabSmartDevices, tabBarVisible: showTabBar(route) })}/>
    </ApplicationTabs.Navigator>
```
- Messages - StackNavigator
```
  <MessageDetails.Navigator
    
    headerMode={"none"}
    screenOptions={Platform.OS === 'android' ? androidStackConfig : stackConfig}

  >
  <MessageDetails.Screen name={RouteIdentifiers.messageDetail.name} component={MessageDetail}/>
  <MessageDetails.Screen name={RouteIdentifiers.survey.name} component={SurveyScreen}/>
  <MessageDetails.Screen name={RouteIdentifiers.safetyFirstDisclaimer.name} component={SafetyFirstDisclaimer}/>
  </MessageDetails.Navigator>
```
- Profile - StackNavigator
```
  <ProfileSettingsNav.Navigator
    
    headerMode={"none"}
    screenOptions={Platform.OS === 'android' ? androidStackConfig : stackConfig}

  >
    <ProfileSettingsNav.Screen name={RouteIdentifiers.profileSettings.name} component = {ProfileSettings}/>
    <ProfileSettingsNav.Screen name={RouteIdentifiers.changePassword.name} component={ChangePassword}/>
    <ProfileSettingsNav.Screen name={RouteIdentifiers.confirmEmail.name} component={ConfirmEmail}/>
    <ProfileSettingsNav.Screen name={RouteIdentifiers.editHomeDetails.name} component={EditHomeDetails}/>
  </ProfileSettingsNav.Navigator>
```
- Inventory - StackNavigator
```
  <InventoryNav.Navigator
    initialRouteName={RouteIdentifiers.inventoryDashboard.name}
    headerMode={"none"}
    screenOptions={Platform.OS === 'android' ? androidStackConfig : stackConfig}
   
  >
    <InventoryNav.Screen name={RouteIdentifiers.inventoryItem.name} component={InventoryItem} />
    <InventoryNav.Screen name={RouteIdentifiers.inventoryItemDetails.name} component={InventoryItemDetails}/>
    <InventoryNav.Screen name={RouteIdentifiers.tag.name} component={Tag}/>
    <InventoryNav.Screen name={RouteIdentifiers.tagInfo.name} component={TagInfo}/>
  </InventoryNav.Navigator>
```

**Your Environment**

| software                       | version |
| ------------------------------ | ------- |
| iOS                 |  14.3
| @react-navigation/native       | 5.9.2
| react-native                   | 0.63

**Comment by Sephaq** [comment]
Yes, issues persist on latest version.

**Comment by satya164** [comment]
https://reactnavigation.org/docs/configuring-links/#rendering-an-initial-route

**Comment by adclark1** [comment]
I had this issue and ended up checking whether you can go back and, if not, sending them to the main app stack that hadn't yet been loaded due to the link/notification by passing the initial load:

`if(navigation.canGoBack()) {
    navigation.goBack();
  }
  else {
    navigation.replace('Tabs');
  }`

**Comment by rrufus** [comment]
I've also come across this issue, which occurs where:
* you are opening the app from a cold start with a deep link (either directly or via a push notification)
* the deep link is meant to correspond to a react navigation screen inside a modal stack

The issue is that the modal page renders without the usual modal "look" where it appears over another screen when opened from a deep link.

Our main issue was that `navigationRef.isReady()` uses a ref (obviously, given the name), so could not be used in a `useEffect` hook. This meant that we could not wait for `isReady` to be true before carrying out the navigation to the modal. 

It seemed that due to this, when the deep link opening the modal was the url to navigate to after a cold start (e.g. from getInitialURL), the modal stack was the only one mounted and so did not render correctly.

We also had a couple of screens which are shown instead of the navigation stack, this seemed to have been an antipattern which introduced bugs in deep link navigation when opening the app from a cold state, due to the navigation not being available.

I achieved this another way following this gist https://gist.github.com/thebuilder/fb07c989093d4a82811625de361884e7#file-hookwithref-js to create:

```
  const [isNavReady, setIsNavReady] = useState(false);

  const setNavigationContainerRef = useCallback(
    (node: NavigationContainerRef<RootStackParamList>) => {
      if (navigationRef.current) {
        // cleanup
      }
      if (node) {
        // add event listeners etc
        node.addListener("state", (state: any) => {
          setIsNavReady(!!state.data);
        });
      }

      navigationRef.current = node;
    },
    []
  );
```

This worked in a way which other logic could be built on top of, which allowed me to fix the bugs I had in navigation.

I also had to alter my `LinkingOptions` object passed as `linking` to the `NavigationContainer` by:
- Disabling `getInitialURL` so that it is a `() => null` and does not use the default, rewriting the logic behind initial routing in a component which used `isNavReady` from above to know when to do the navigation.
- Using the default `subscribe` and rewriting the logic to handle routing from push notifications in another component where it can use `isNavReady` to know when to do the navigation.

Just the way that I ended up solving this issue for my company, but open to suggestions.

A larger refactor which may have been difficult for other reasons could've included removing all logic which rendered over/instead-of navigation stacks, and replacing those with screens. Then, ensuring that users cannot navigate away from these screens to achieve the same effect that was initially there (eg for a paywall).
